src/tree.rs

```diff
@@ -1423,6 +1423,81 @@ impl<T> Tree<T> {
     }
 }
 
+impl<T> PartialEq for Tree<T>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &Tree<T>) -> bool {
+        if self.nodes.iter().filter(|x| x.is_some()).count()
+            != other.nodes.iter().filter(|x| x.is_some()).count()
+        {
+            return false;
+        }
+
+        for (node1, node2) in self
+            .nodes
+            .iter()
+            .filter_map(|x| (*x).as_ref())
+            .zip(other.nodes.iter().filter_map(|x| (*x).as_ref()))
```

Comment:
Question: do you know why we have a vector of Option for nodes in Tree instead of just having a vector of Node? I'm questioning all the logic about the free_ids too here.

We won't need the .filter_map(|x| (*x).as_ref()) if we had a Vec<Node> instead of Vec<Option<Node>